cirq/experiments/fidelity_estimation.py

```diff
@@ -534,35 +532,67 @@ def sample_2q_xeb_circuits(
     return pd.DataFrame(records).set_index(['circuit_i', 'cycle_depth'])
 
 
-def _simulate_2q_xeb_circuit(task: Dict[str, Any]):
-    """Helper function for simulating a given (circuit, cycle_depth)."""
-    circuit_i = task['circuit_i']
-    cycle_depth = task['cycle_depth']
-    circuit = task['circuit']
-    param_resolver = task['param_resolver']
+@dataclass(frozen=True)
+class _Simulate2qXEBTask:
+    """Helper container for executing simulation tasks, potentially via multiprocessing."""
+
+    circuit_i: int
+    cycle_depths: Sequence[int]
+    circuit: 'cirq.Circuit'
+    param_resolver: 'cirq.ParamResolverOrSimilarType'
 
-    circuit_depth = cycle_depth * 2 + 1
-    assert circuit_depth <= len(circuit)
-    tcircuit = circuit[:circuit_depth]
-    tcircuit = protocols.resolve_parameters_once(tcircuit, param_resolver=param_resolver)
 
-    pure_sim = sim.Simulator()
-    psi = cast(sim.StateVectorTrialResult, pure_sim.simulate(tcircuit))
-    psi = psi.final_state_vector
-    pure_probs = np.abs(psi) ** 2
+class _Simulate_2q_XEB_Circuit:
+    """Closure used in `simulate_2q_xeb_circuits` so it works with multiprocessing."""
 
-    return {
-        'circuit_i': circuit_i,
-        'cycle_depth': cycle_depth,
-        'pure_probs': pure_probs,
-    }
+    def __init__(self, simulator):
+        self.simulator = simulator
+
+    def __call__(self, task: _Simulate2qXEBTask):
+        """Helper function for simulating a given (circuit, cycle_depth)."""
+        circuit_i = task.circuit_i
+        cycle_depths = set(task.cycle_depths)
+        circuit = task.circuit
+        param_resolver = task.param_resolver
+
+        records = []
+        for moment_i, step_result in enumerate(
+            self.simulator.simulate_moment_steps(circuit=circuit, param_resolver=param_resolver)
+        ):
+            # circuit_depth = cycle_depth * 2 + 1
+            # step_result is the result *after* moment_i, so
+            # circuit_depth = moment_i + 1
+            # moment_i = cycle_depth * 2
+            if moment_i % 2 == 1:
+                continue
+            cycle_depth = moment_i // 2
+            if cycle_depth not in cycle_depths:
+                continue
+            cycle_depths.remove(cycle_depth)
+
+            psi = cast(sim.SparseSimulatorStep, step_result)
```

Comment:
Usually I would factor out functionality to reduce code duplication. Otherwise, I find the "linear" version easier to read, and it's especially easier to deduce the record keys of this function without having to scroll to yet-another-helper-function.